ui-modules/app-inspector/app/components/entity-tree/entity-tree.directive.js

```diff
@@ -141,55 +134,106 @@ export function entityTreeDirective() {
+                            // view mode, only if it does not play the role of 'OTHER_CHILD' at the same time.
+                            if (!childRole) {
+                                displayEntityInView(entity, view);
+                                highlightEntityInView(entity, view);
+                            }
+
+                            // Look for 'OTHER_CHILD' entities under 'OTHER_PARENT', flip or move them and label to display in
+                            // 'OTHER_PARENT/OTHER_CHILD' view mode respectively.
+                            parentRole.targets.forEach(target => {
+                                let relatedEntity = findEntity(entities, target);
+                                if (relatedEntity) {
+                                    highlightEntityInView(relatedEntity, view);
+                                    displayParentsInView(entities, relatedEntity.parentId, view);
+
+                                    // Re-arrange the tree if related 'OTHER_CHILD' entity is not a child of 'OTHER_PARENT'.
+                                    if (relatedEntity.parentId !== entity.id) {
+
+                                        if (relatedEntity.id === entity.parentId) {
+                                            // 4.1. Flip 'OTHER_CHILD' parent with a 'OTHER_PARENT' child.
+                                            flipParentAndChild(relatedEntity, entity, entities, view);
+                                        } else {
+                                            // 4.2. Move 'OTHER_CHILD' entity to a new 'OTHER_PARENT' parent.
+                                            moveEntityToParent(relatedEntity, entity, entities, view);
+                                        }
                                     }
                                 }
-                            }
-                        });
-                    } else if (!relationship || relationship.name !== RELATIONSHIP_HOSTED_ON) {
+                            });
+                        } else if (!parentRole && !childRole) {
 
-                        // Display original position for any other entity under 'host_for/hosted_on' view. Do no highlight
-                        // entities that are required to be displayed but do not belong to this view.
-                        displayEntityInView(entity, VIEW_HOST_FOR_HOSTED_ON);
-                    }
+                            // Display original position for any other entity under 'OTHER_PARENT/OTHER_CHILD' view. Do
+                            // no highlight entities that are required to be displayed but do not belong to this view.
+                            displayEntityInView(entity, view);
+                        }
+                    });
+
+                    // Phase 2. Look through all entities found again and process entities that play both roles:
+                    //          'OTHER_PARENT' and 'OTHER_CHILD'.
```

Comment:
New phase to cover edge case when entity plays role of parent and child in the view.